verification/src/http_server/handlers/verification/sourcify/metadata.rs

```diff
@@ -0,0 +1,187 @@
+use std::collections::BTreeMap;
+
+use ethers_solc::EvmVersion;
+use serde::Deserialize;
+
+use crate::VerificationResult;
+
+use super::types::Files;
+
+const METADATA_FILE_NAME: &str = "metadata.json";
+
+#[derive(Debug, PartialEq, Deserialize)]
+
+pub struct Metadata {
+    pub settings: MetadataSettings,
+    pub compiler: Compiler,
+    pub output: Output,
+}
+
+#[derive(Debug, PartialEq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MetadataSettings {
+    pub compilation_target: BTreeMap<String, String>,
+    pub optimizer: Optimizer,
+    pub libraries: BTreeMap<String, String>,
+    #[serde(rename = "camelCase")]
+    pub evm_version: Option<EvmVersion>,
+}
+
+#[derive(Debug, PartialEq, Deserialize)]
+pub struct Optimizer {
+    pub enabled: Option<bool>,
+    pub runs: Option<usize>,
+}
+
+#[derive(Debug, PartialEq, Deserialize)]
+pub struct Compiler {
+    pub version: String,
+}
+
+#[derive(Debug, PartialEq, Deserialize)]
+pub struct Output {
+    pub abi: serde_json::Value,
+}
+
+impl TryFrom<&Files> for Metadata {
+    type Error = anyhow::Error;
+
+    fn try_from(files: &Files) -> Result<Self, Self::Error> {
+        let metadata_content = files
+            .0
+            .get(METADATA_FILE_NAME)
+            .ok_or_else(|| anyhow::Error::msg(format!("file {} not found", METADATA_FILE_NAME)))?;
+
+        serde_json::from_str(metadata_content.as_str()).map_err(anyhow::Error::msg)
+    }
+}
+
+impl TryFrom<Files> for VerificationResult {
+    type Error = anyhow::Error;
+
+    fn try_from(files: Files) -> Result<Self, Self::Error> {
+        let metadata = Metadata::try_from(&files)?;
+        let contract_name = metadata
+            .settings
+            .compilation_target
+            .iter()
+            .next()
+            .ok_or_else(|| anyhow::Error::msg("compilation target not found"))?
+            .1
+            .to_string();
+        let compiler_version = metadata.compiler.version;
+        let evm_version = metadata
+            .settings
+            .evm_version
+            .unwrap_or_default()
+            .to_string();
+        let optimization_runs = metadata.settings.optimizer.enabled.and_then(|enabled| {
```

Comment:
fyi it is possible to have "enabled: false" and it will still do some optimizations (leaving enabled to nil and enabled:false is two different options, sadly)
I'll try to make a PR with this idea in mind